### PR TITLE
Implement `Extractlane`, `UaddSat`, and `UsubSat` for Cranelift interpreter

### PR DESCRIPTION
**Pull request by dheaton-arm**
Implemented the `Extractlane`, `UaddSat`, and `UsubSat` opcodes for the interpreter,
and added helper functions for working with SIMD vectors (`extractlanes`, `vectorizelanes`,
and `binary_arith`).

Copyright (c) 2021, Arm Limited

### CODE REVIEW

**Review by afonso360 — CHANGES_REQUESTED** [review]

File: cranelift/interpreter/src/value.rs
Comment (abrown) [inline cranelift/interpreter/src/value.rs:234]: This might need to be more strict:
```suggestion
        assert!(ty.is_vector() && elem_ty.bytes() == 16);
```

File: cranelift/interpreter/src/value.rs
Comment (abrown) [inline cranelift/interpreter/src/value.rs:389]: What do you think about implementing these instructions as: convert to larger type, use already existing add/sub, convert back to the right type with saturation (maybe with a new saturating `ValueConversionKind`). I can't remember how many instructions end up being saturating ones but if there are more it could reduce the size of `DataValue`.

Comment (dheaton-arm) [reply cranelift/interpreter/src/value.rs:389]: Looking at the list of opcodes, there's a few but given that a lot of them are signed/unsigned variants of each other (which I don't think will need re-implementing each time, as in this case for example), I think there'd be ~5 saturating ops in DataValue. We could do that suggestion though if preferred, which might make implementing opcodes like the `narrow` set easier later (the only question is if `[I|U]128` is supposed to be supported; I see a few of the other functions in DataValue seem to ignore it and being the typically largest integer type would require different implementation).

File: cranelift/filetests/filetests/interpreter/extractlane.clif
Comment (dheaton-arm) [reply cranelift/filetests/filetests/interpreter/extractlane.clif:1]: Trying this I'm getting errors that `i32x4` is an unsupported type for `uadd_sat` and `usub_sat` on the `test run` side -- should that be the case?

File: cranelift/interpreter/src/step.rs
Comment (afonso360) [inline cranelift/interpreter/src/step.rs:867]: Would it be easier to do a blanket impl of the `Value` trait like: `impl<T: Value, const N: usize> Value for [T; N]`? And have a function that would convert the SIMD values into/from the array?

I don't know how many of these functions we are going to need, but it might be worth considering in the future.

Although I think this would require rust 1.51 with `min_const_generics`. Which might be higher than our minimum requirements.

File: cranelift/interpreter/src/step.rs
Comment (afonso360) [inline cranelift/interpreter/src/step.rs:879]:
```suggestion
    for (lhs, rhs) in arg0.iter().zip(arg1.iter()) {
        // The initial Value::int needs to be on a separate line so the
        // compiler can determine concrete types.
        let mut lhs: V = Value::int(*lhs, vector_type.lane_type())?;
        let mut rhs: V = Value::int(*rhs, vector_type.lane_type())?;
```

File: cranelift/filetests/filetests/interpreter/extractlane.clif
Comment (afonso360) [reply cranelift/filetests/filetests/interpreter/extractlane.clif:1]: Hmm, according to the [docs](https://docs.rs/cranelift-codegen/0.76.0/cranelift_codegen/ir/trait.InstBuilder.html#method.uadd_sat), it should work.

However, the reason that it probably doesn't work is that its not included in the [WASM SIMD Spec](https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#saturating-integer-addition).

What we usually do in these cases is separate the testcases in two files. `8x16` / `16x8` in one file with the run targets, and `i32x4` / `i64x2` (could you add testcases for `i64x2` as well?) into another with just the interpreter.

I also usually file an issue stating that these are not implemented in the x86_64 and aarch64 backends. We've been finding a lot of code that should compile but doesn't since we started adding these tests instead of relying on the wasm test suite.

**Comment by afonso360** [comment]
Thanks for filing the issue!

Could you add a `simd-` prefix to the test file names? Just to keep in line with the other tests on that directory.

Otherwise LGTM! Thank you!

**Review by afonso360 — APPROVED** [review]

File: cranelift/interpreter/src/step.rs
Comment (cfallin) [inline cranelift/interpreter/src/step.rs:817]: I think it might be slightly preferable here to use `SmallVec`, perhaps `SmallVec<[i128; 1]>` or similar (and typedef it so we can easily change it) -- while we don't expect the interpreter to be *fast*, it's still good to avoid an unnecessary heap allocation per execution of an instruction.

File: cranelift/interpreter/src/value.rs
Comment (abrown) [reply cranelift/interpreter/src/value.rs:389]: Yeah, it sounds like the current approach will be fine.

Comment (abrown) [reply cranelift/interpreter/src/value.rs:389]: Thanks for adding these instructions!

File: cranelift/interpreter/src/step.rs
Comment (dheaton-arm) [reply cranelift/interpreter/src/step.rs:817]: Is this specifically for `extractlanes` or would this also apply to the input to `vectorizelanes`? Just want to check in case you wanted basically all the `Vec`s in this section of code changing, whilst I'm thinking that parameters aren't an issue (which I may be wrong about) and `Vec` parameters seem more general / flexible (accepting both `SmallVec` and `Vec`).

Comment (cfallin) [reply cranelift/interpreter/src/step.rs]: I guess the advice applies generally. Specifically, this:

> `Vec` parameters seem more general / flexible (accepting both `SmallVec` and `Vec`)

is not quite right; `Vec` is a specific type that requires its contents to be heap-allocated. The more idiomatic Rust type for "any array of values in memory" would be a slice, i.e. `&[128]`, which one can get with `&vec[..]` or `&smallvec[..]`; but that only works for inbound args because it has no ownership. Returning a `SmallVec` is a technique we use elsewhere pretty successfully so I think it's' reasonable here too :-)

Comment (dheaton-arm) [reply cranelift/interpreter/src/step.rs]: Ah, I was basing that statement off testing where using a `Vec` parameter but passing a `SmallVec` was accepted by the compiler and built successfully, but the inverse (using a `SmallVec` parameter but passing a `Vec`) was blocked by the compiler. I wasn't sure about if that made the `SmallVec` behave like a `Vec` with regards to heap allocations, however, hence my question.
> Returning a SmallVec is a technique we use elsewhere pretty successfully

I understood that, and changed it to behave like that in 0089e15; I was more wondering about the input parameters to `vectorizelanes`. I'm gathering from your response though that the preference is for those to be a slice?

**Review by cfallin — APPROVED** [review]
Thanks!

**Comment by cfallin** [comment]
There's a cargo-deny issue that I think you'll need to rebase over #3227 to fix; once you've done that, then I'm happy to merge.

**Pull request merged**
